### PR TITLE
#7177: add knip config for dead-code/export/import detection

### PR DESCRIPTION
**Pull request by twschiller**
## What does this PR do?

- Part of #7177
- Adds a knip configuration
- Adds knip to CI with the following rules set to errors:
  - duplicate errors
- Eliminates duplicate exports from the project
- Drops some unused dependencies
- Removes some unused files from the project

## Reviewer Tips

- Review knip.js
- Review dropped dead code

## Demo

```
npm run dead-code
```

image

## Future Work

- Incrementally work through configuration, unused dependencies, false positives to enforce more rules during CI

## Checklist

- [x] Add tests: N/A
- [x] New files added to `src/tsconfig.strictNullChecks.json` (if possible): dropped files that no longer exist
- [x] Designate a primary reviewer: @grahamlangford

### CODE REVIEW

File: knip.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    types: "warn",
+    nsTypes: "warn",
+    enumMembers: "warn",
+    classMembers: "warn",
+    // Incrementally enforce rules over time
+    duplicates: "error",
+  },
+};
+
+console.log(JSON.stringify(knipConfig, null, 2));
```
Comment (grahamlangford) [inline knip.js:49]: Was this left in intentionally?

File: knip.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
+const configFactory = require("./webpack.config.js");
+
+const config = configFactory(process.env, {});
+
+// https://knip.dev/overview/configuration#customize
+const knipConfig = {
+  $schema: "https://unpkg.com/knip@3/schema.json",
+  // Try to get webpack plugin working
```
Comment (grahamlangford) [inline knip.js:8]: Is it not working? What's the problem?

File: src/hooks/useRenderCount.ts
Comment (grahamlangford) [inline src/hooks/useRenderCount.ts:1]: This is a test utility I was keeping around for performance analysis. (Used to to detect some of the performance issues we were experiencing in both the Page Editor and the Sidebar).

**Review by grahamlangford — APPROVED** [review]

File: src/hooks/useRenderCount.ts
Comment (twschiller) [reply src/hooks/useRenderCount.ts:1]: 👍  I moved to the `src/development/` folder and marked the folder as excluded for knip

File: src/utils/inference/selectorInference.ts [hunk 3]
```diff
@@ -38,9 +38,8 @@ import { type ElementInfo } from "@/utils/inference/selectorTypes";
 import { getAttributeSelectorRegex } from "@/utils/inference/selectorInferenceUtils";
 import { assertNotNullish } from "@/utils/nullishUtils";
 
-export const NON_EXISTENT_TAG_NAME = "selectnothing";
 /** Valid selector that never returns any element (no `<selectnothing>` element exists) */
-export const SELECTOR_THAT_MATCHES_NOTHING = NON_EXISTENT_TAG_NAME;
```
Comment (fregante) [inline src/utils/inference/selectorInference.ts:43]: Good, it was a leftover of previous selector tries

File: src/development/hooks/useRenderCount.ts [hunk 4]
```diff
@@ -17,10 +17,15 @@
...
+ */
+function useRenderCount() {
   const count = useRef(0);
 
   count.current += 1;
 
   return count.current;
 }
+
+export default useRenderCount;
```
Comment (fregante) [inline src/development/hooks/useRenderCount.ts:31]: Was this change forced by the tool?

**Review by fregante — APPROVED** [review]
Big fan, I'm surprised it dropped only these

File: src/development/hooks/useRenderCount.ts (hunk 4 above)
Comment (twschiller) [reply src/development/hooks/useRenderCount.ts:31]: No - personal preference. Could use a lint rule to enforce a particular style

**Pull request merged**